### PR TITLE
🚑 Fix scitypes for all existing methods

### LINKED ISSUE #76: On `input_scitype` for static models
The meaning of `input_scitype` for `Static` models is not clear in the MLJ spec, a fact this is corrected [here](https://github.com/alan-turing-institute/MLJ.jl/blob/static-docs/docs/src/adding_models_for_general_use.md#static-models-models-that-do-not-generalize). My apologies for not catching this earlier, but now this is clarified, the implementation here is no longer compliant.

In the current package the `input_scitype` should refer to a bound on the `scitpe((X, y))` for `X` and `y` in `transform(model, _, X, y)`. So, for `SMOTE`, for example, I believe this should be

```
Tuple{
    Union{
        Table(Continuous), 
        AbstractMatrix{Continuous}},
    },
    AbstractVector
}
```

whereas it is currently 

```julia
Union{
    Table(Continuous),
    AbstractMatrix{Continuous}},
},
```

While fixing this is technically breaking, I think we can guarantee it actually breaks no existing code and can regard this a  bug fix.

### PR DESCRIPTION
**Pull request by EssamWisam**
Closes #76

### CODE REVIEW

**Comment by EssamWisam** [comment]
@ablaom 
This maps the old setup into one where the scitypes are correct as needed for #76.

**Comment by EssamWisam** [comment]
Regarding the last commit, I noticed that a change I have done late in the development of the package to let random resampling operate on categorical types was not reflected on the scitypes used. The commit fix this. In this, I also assumed that if a matrix is `AbstractMatrix{<:Real}` then it also is `AbstractMatrix{Infinite}`

File: src/oversampling_methods/random_walk/interface_mlj.jl [hunk 1]
```diff
@@ -46,11 +46,12 @@ MMI.metadata_model(
...
         Table(Union{Infinite, Finite}),
+        AbstractVector
     },
     output_scitype = Union{
         Table(Union{Infinite, Finite}),
+        AbstractVector
     },
-    target_scitype = AbstractVector,
     load_path = "Imbalance.MLJ.RandomWalkOversampler"
 )
```
Comment (ablaom) [inline src/oversampling_methods/random_walk/interface_mlj.jl:56]: Why is this one not repaired in the same pattern as the others? Don't you need `Tuple` somewhere?

File: src/oversampling_methods/smotenc/interface_mlj.jl [hunk 2]
```diff
@@ -61,11 +61,12 @@ MMI.metadata_model(
     SMOTENC,
     input_scitype = Union{
         Table(Union{Infinite, Finite}),
+        AbstractVector
     },
     output_scitype = Union{
         Table(Union{Infinite, Finite}),
+        AbstractVector
     },
```
Comment (ablaom) [inline src/oversampling_methods/smotenc/interface_mlj.jl:69]: Why is this one not repaired in the same pattern as the others? Don't you need `Tuple` somewhere?

File: src/oversampling_methods/borderline_smote1/interface_mlj.jl [hunk 3]
```diff
@@ -65,9 +65,20 @@ MMI.metadata_pkg(
...
+                        }, 
+                        AbstractVector
+                    },
+    output_scitype = Tuple{
+        Union{
+            Table(Continuous),
+            AbstractMatrix{Continuous}
+        }, 
+        AbstractVector
+    },
```
Comment (ablaom) [inline src/oversampling_methods/borderline_smote1/interface_mlj.jl:81]: Whitespace suggestion:

```suggestion
    input_scitype = Tuple{
         Union{
            Table(Continuous),
            AbstractMatrix{Continuous}
        }, 
        AbstractVector
    },
    output_scitype = Tuple{
        Union{
            Table(Continuous),
            AbstractMatrix{Continuous}
        }, 
        AbstractVector
    },
```

**Comment by ablaom** [comment]
>  The commit fix this. In this, I also assumed that if a matrix is AbstractMatrix{<:Real} then it also is AbstractMatrix{Infinite}

Strictly speaking this is not correct because `Rational <: Real` and `AbstractIrrational <: Real` but objects subtyping these have `Unknown` scitype. On reflection, we could perhaps have included them in the `Continuous` scitype. (We couldn't include all `Real` because `Integer <: Real` and `Integer` is for `Count`. )

File: src/oversampling_methods/random_walk/interface_mlj.jl (hunk 1 above)
Comment (EssamWisam) [reply src/oversampling_methods/random_walk/interface_mlj.jl:56]: Oops! I did this shortly before sleeping and looks like I took Union{ to be Tuple{ as it was indented similar to other methods.

**Comment by EssamWisam** [comment]
> On reflection, we could perhaps have included them in the `Continuous` scitype. (We couldn't include all `Real` because `Integer <: Real` and `Integer` is for `Count`. )

Sure, I could do that but then a matrix of integers (corresponding to a purely categorical dataset) would work perfectly with the functional interface but not the MLJ interface; meanwhile, it's quite odd to have a dataset of irrational numbers or numbers taking the rational form a//b.

What do you think?

**Review by ablaom — APPROVED** [review]
Looks good to me, thank you.

**Pull request merged**